### PR TITLE
Support declarative extension registration on fields and parameters

### PR DESCRIPTION
**Pull request by sbrannen**
## Overview

This is a PR for #864.

The implementation is essentially feature complete and tested (following the 80/20 rule); however, there are some points that should be discussed before merging this PR into the `main` branch.

## Topics for Discussion

When discussing any of these topics, I recommend that you review the current behavior by analyzing the tests in `ExtensionRegistrationViaParametersAndFieldsTests` before spending time studying the actual implementation details.

1. **Registration Order**: I have done my best to register extensions in the order in which the fields or parameters are used in the lifecycle of a test class and test method. This can be inspected in the `registrationOrder()` test method. Implementation details can be found in `ClassBasedTestDescriptor` and `TestMethodTestDescriptor`.
2. **`@Order` Support**: `ExtensionUtils.registerExtensionsFromFields()` already supported `@Order` for extensions registered via `@RegisterExtension`, and `@Order` is now also applied to extensions registered via fields using `@ExtendWith`. But... `@ExtendWith` extensions are registered _before_ `@RegisterExtension` extensions.
3. **`@ExtendWith` target**: `@ExtendWith` can currently be declared on a `TYPE` or `METHOD`. It cannot be declared directly on a `PARAMETER` or `FIELD`.
4. **Private fields**: `@RegisterExtension` fields are not allowed to be `private`, but in this PR fields meta-annotated with `@ExtendWith` are allowed to be `private`.

I will share my thoughts on these topics as comments and invite feedback.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sbrannen** [comment]
> **Private fields**: `@RegisterExtension` fields are not allowed to be `private`, but in this PR fields meta-annotated with `@ExtendWith` are allowed to be `private`.

From one perspective, this is inconsistent treatment of fields; however, I wonder if we haven't been too strict with regard to the current non-private policy for `@RegisterExtension` fields.

If JUnit or a third party provides an extension that is implicitly registered by a custom composed annotation (which is meta-annotated with `@ExtendWith`), then it is really up to the user of that composed annotation to decide if the annotated field is `private`.

As long as the extension can work with a `private` field, there should be no unnecessary constraint on the visibility of the field.

In other words, why should JUnit not allow the user to decide if a field is `private`?

I recommend that we allow `@RegisterExtension` and `@ExtendWith` fields to be `private`.

**Comment by sbrannen** [comment]
> **`@ExtendWith` target**: `@ExtendWith` can currently be declared on a `TYPE` or `METHOD`. It cannot be declared directly on a `PARAMETER` or `FIELD`.

Although the original intention of #864 was to allow extension authors to provide custom composed annotations that are meta-annotated with `@ExtendWith`, after having spent more time with the issue I think there are likely valid use cases for allowing `@ExtendWith` to be declared directly on a field or parameter.

In other words, I don't believe we should force users or library implementers to create custom composed annotations when all they need is `@ExtendWith` on a field or parameter.

**Comment by sbrannen** [comment]
> **`@Order` Support**: `ExtensionUtils.registerExtensionsFromFields()` already supported `@Order` for extensions registered via `@RegisterExtension`, and `@Order` is now also applied to extensions registered via fields using `@ExtendWith`. But... `@ExtendWith` extensions are registered _before_ `@RegisterExtension` extensions.

The above describes the status quo in this PR, but we can certainly change the semantics.

My rationale for implementing the feature with these semantics was based on the fact that extensions registered via `@ExtendWith` on a class or interface are always registered before any extensions registered via `@RegisterExtension`. This is the current behavior in JUnit Jupiter, and that default behavior will not change.

So, the status quo in this PR is consistent in that regard.

However, we have an opportunity (for the first time) to make it possible for extensions registered via `@ExtendWith` to be ordered by the user **relative** to extensions registered via `@RegisterExtension`.

Achieving that is straight forward: I'd just need to collapse the two iterations in `ExtensionUtils.registerExtensionsFromFields()` into a single iteration over the fields.

After having spent more time with this issue, I am in favor of allowing `@RegisterExtension` and `@ExtendWith` fields to be ordered relative to each other, and I imagine the community would likely expect all such fields to be sorted together when using `@Order`.

**Comment by sbrannen** [comment]
> **Registration Order**: I have done my best to register extensions in the order in which the fields or parameters are used in the lifecycle of a test class and test method. This can be inspected in the `registrationOrder()` test method. Implementation details can be found in `ClassBasedTestDescriptor` and `TestMethodTestDescriptor`.

Aside from the discussion about `@Order` support, I feel pretty comfortable with the current extension registration order, but feedback is certainly welcome!

**Comment by sbrannen** [comment]
> However, we have an opportunity (for the first time) to make it possible for extensions registered via `@ExtendWith` to be ordered by the user **relative** to extensions registered via `@RegisterExtension`.

Related issues:

- #506 
- #2651

**Comment by sbrannen** [comment]
**Team Decision**:

- Allow `@RegisterExtension` and `@ExtendWith` fields to be `private`.
- Allow `@ExtendWith` to be declared directly on fields and parameters.
- Allow `@RegisterExtension` and `@ExtendWith` fields to be sorted via `@Order` relative to each other.
- Throw an exception if a field is annotated with both `@RegisterExtension` and `@ExtendWith`.

**Review by marcphilipp — APPROVED** [review]

**Comment by sbrannen** [comment]
Closed via commit 8b5387c5bef050730d8feb52a5292a32530f65cd
